[Feature]: Improve GGUF loading from HuggingFace user experience like repo_id:quant_type

injaeryou opened the pull request:
## Purpose
Fixes #25182
Improve GGUF loading from HuggingFace user experience like repo_id:quant_type

```
vllm serve unsloth/Qwen3-0.6B-GGUF:IQ1_S --tokenizer Qwen/Qwen3-0.6B
```

## Test Plan
```
vllm serve unsloth/Qwen3-0.6B-GGUF:IQ1_S --tokenizer Qwen/Qwen3-0.6B
pytest tests/models/test_gguf_download.py
pytest tests/transformers_utils/test_utils.py
```

## Test Result
<details>
<summary> [BEFORE] vllm serve unsloth/Qwen3-0.6B-GGUF:IQ1_S --tokenizer Qwen/Qwen3-0.6B </summary>

```
 vllm serve unsloth/Qwen3-0.6B-GGUF:IQ1_S --tokenizer Qwen/Qwen3-0.6B
INFO 11-21 10:31:03 [scheduler.py:207] Chunked prefill is enabled with max_num_batched_tokens=2048.
(APIServer pid=1210609) INFO 11-21 10:31:03 [api_server.py:1978] vLLM API server version 0.11.2.dev67+g1d642872a.d20251120
(APIServer pid=1210609) INFO 11-21 10:31:03 [utils.py:253] non-default args: {'model_tag': 'unsloth/Qwen3-0.6B-GGUF:IQ1_S', 'model': 'unsloth/Qwen3-0.6B-GGUF:IQ1_S', 'tokenizer': 'Qwen/Qwen3-0.6B'}
(APIServer pid=1210609) Traceback (most recent call last):
(APIServer pid=1210609)   File "/home/name/.test/vllm/.venv/lib/python3.12/site-packages/transformers/utils/hub.py", line 479, in cached_files
(APIServer pid=1210609)     hf_hub_download(
(APIServer pid=1210609)   File "/home/name/.test/vllm/.venv/lib/python3.12/site-packages/huggingface_hub/utils/_validators.py", line 106, in _inner_fn
(APIServer pid=1210609)     validate_repo_id(arg_value)
(APIServer pid=1210609)   File "/home/name/.test/vllm/.venv/lib/python3.12/site-packages/huggingface_hub/utils/_validators.py", line 160, in validate_repo_id
(APIServer pid=1210609)     raise HFValidationError(
(APIServer pid=1210609) huggingface_hub.errors.HFValidationError: Repo id must use alphanumeric chars, '-', '_' or '.'. The name cannot start or end with '-' or '.' and the maximum length is 96: 'unsloth/Qwen3-0.6B-GGUF:IQ1_S'.
(APIServer pid=1210609)
(APIServer pid=1210609) During handling of the above exception, another exception occurred:
(APIServer pid=1210609)
(APIServer pid=1210609) Traceback (most recent call last):
(APIServer pid=1210609)   File "/home/name/.test/vllm/.venv/lib/python3.12/site-packages/transformers/configuration_utils.py", line 721, in _get_config_dict
(APIServer pid=1210609)     resolved_config_file = cached_file(
(APIServer pid=1210609)                            ^^^^^^^^^^^^
(APIServer pid=1210609)   File "/home/name/.test/vllm/.venv/lib/python3.12/site-packages/transformers/utils/hub.py", line 322, in cached_file
(APIServer pid=1210609)     file = cached_files(path_or_repo_id=path_or_repo_id, filenames=[filename], **kwargs)
(APIServer pid=1210609)            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(APIServer pid=1210609)   File "/home/name/.test/vllm/.venv/lib/python3.12/site-packages/transformers/utils/hub.py", line 532, in cached_files
(APIServer pid=1210609)     _get_cache_file_to_return(path_or_repo_id, filename, cache_dir, revision, repo_type)
(APIServer pid=1210609)   File "/home/name/.test/vllm/.venv/lib/python3.12/site-packages/transformers/utils/hub.py", line 143, in _get_cache_file_to_return
(APIServer pid=1210609)     resolved_file = try_to_load_from_cache(
(APIServer pid=1210609)                     ^^^^^^^^^^^^^^^^^^^^^^^
(APIServer pid=1210609)   File "/home/name/.test/vllm/.venv/lib/python3.12/site-packages/huggingface_hub/utils/_validators.py", line 106, in _inner_fn
(APIServer pid=1210609)     validate_repo_id(arg_value)
(APIServer pid=1210609)   File "/home/name/.test/vllm/.venv/lib/python3.12/site-packages/huggingface_hub/utils/_validators.py", line 160, in validate_repo_id
(APIServer pid=1210609)     raise HFValidationError(
(APIServer pid=1210609) huggingface_hub.errors.HFValidationError: Repo id must use alphanumeric chars, '-', '_' or '.'. The name cannot start or end with '-' or '.' and the maximum length is 96: 'unsloth/Qwen3-0.6B-GGUF:IQ1_S'.
(APIServer pid=1210609)
(APIServer pid=1210609) During handling of the above exception, another exception occurred:
(APIServer pid=1210609)
(APIServer pid=1210609) Traceback (most recent call last):
(APIServer pid=1210609)   File "/home/name/.test/vllm/.venv/bin/vllm", line 10, in <module>
(APIServer pid=1210609)     sys.exit(main())
(APIServer pid=1210609)              ^^^^^^
(APIServer pid=1210609)   File "/home/name/.test/vllm/vllm/entrypoints/cli/main.py", line 73, in main
(APIServer pid=1210609)     args.dispatch_function(args)
(APIServer pid=1210609)   File "/home/name/.test/vllm/vllm/entrypoints/cli/serve.py", line 60, in cmd
(APIServer pid=1210609)     uvloop.run(run_server(args))
(APIServer pid=1210609)   File "/home/name/.test/vllm/.venv/lib/python3.12/site-packages/uvloop/__init__.py", line 96, in run
(APIServer pid=1210609)     return __asyncio.run(
(APIServer pid=1210609)            ^^^^^^^^^^^^^^
(APIServer pid=1210609)   File "/usr/lib/python3.12/asyncio/runners.py", line 194, in run
(APIServer pid=1210609)     return runner.run(main)
(APIServer pid=1210609)            ^^^^^^^^^^^^^^^^
(APIServer pid=1210609)   File "/usr/lib/python3.12/asyncio/runners.py", line 118, in run
(APIServer pid=1210609)     return self._loop.run_until_complete(task)
(APIServer pid=1210609)            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(APIServer pid=1210609)   File "uvloop/loop.pyx", line 1518, in uvloop.loop.Loop.run_until_complete
(APIServer pid=1210609)   File "/home/name/.test/vllm/.venv/lib/python3.12/site-packages/uvloop/__init__.py", line 48, in wrapper
(APIServer pid=1210609)     return await main
(APIServer pid=1210609)            ^^^^^^^^^^
(APIServer pid=1210609)   File "/home/name/.test/vllm/vllm/entrypoints/openai/api_server.py", line 2028, in run_server
(APIServer pid=1210609)     await run_server_worker(listen_address, sock, args, **uvicorn_kwargs)
(APIServer pid=1210609)   File "/home/name/.test/vllm/vllm/entrypoints/openai/api_server.py", line 2047, in run_server_worker
(APIServer pid=1210609)     async with build_async_engine_client(
(APIServer pid=1210609)   File "/usr/lib/python3.12/contextlib.py", line 210, in __aenter__
(APIServer pid=1210609)     return await anext(self.gen)
(APIServer pid=1210609)            ^^^^^^^^^^^^^^^^^^^^^
(APIServer pid=1210609)   File "/home/name/.test/vllm/vllm/entrypoints/openai/api_server.py", line 196, in build_async_engine_client
(APIServer pid=1210609)     async with build_async_engine_client_from_engine_args(
(APIServer pid=1210609)   File "/usr/lib/python3.12/contextlib.py", line 210, in __aenter__
(APIServer pid=1210609)     return await anext(self.gen)
(APIServer pid=1210609)            ^^^^^^^^^^^^^^^^^^^^^
(APIServer pid=1210609)   File "/home/name/.test/vllm/vllm/entrypoints/openai/api_server.py", line 222, in build_async_engine_client_from_engine_args
(APIServer pid=1210609)     vllm_config = engine_args.create_engine_config(usage_context=usage_context)
(APIServer pid=1210609)                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(APIServer pid=1210609)   File "/home/name/.test/vllm/vllm/engine/arg_utils.py", line 1358, in create_engine_config
(APIServer pid=1210609)     maybe_override_with_speculators(
(APIServer pid=1210609)   File "/home/name/.test/vllm/vllm/transformers_utils/config.py", line 573, in maybe_override_with_speculators
(APIServer pid=1210609)     config_dict, _ = PretrainedConfig.get_config_dict(
(APIServer pid=1210609)                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(APIServer pid=1210609)   File "/home/name/.test/vllm/.venv/lib/python3.12/site-packages/transformers/configuration_utils.py", line 662, in get_config_dict
(APIServer pid=1210609)     config_dict, kwargs = cls._get_config_dict(pretrained_model_name_or_path, **kwargs)
(APIServer pid=1210609)                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
(APIServer pid=1210609)   File "/home/name/.test/vllm/.venv/lib/python3.12/site-packages/transformers/configuration_utils.py", line 744, in _get_config_dict
(APIServer pid=1210609)     raise OSError(
(APIServer pid=1210609) OSError: Can't load the configuration of 'unsloth/Qwen3-0.6B-GGUF:IQ1_S'. If you were trying to load it from 'https://huggingface.co/models', make sure you don't have a local directory with the same name. Otherwise, make sure 'unsloth/Qwen3-0.6B-GGUF:IQ1_S' is the correct path to a directory containing a config.json file
```
</details>

<details>
<summary> [AFTER] vllm serve unsloth/Qwen3-0.6B-GGUF:IQ1_S --tokenizer Qwen/Qwen3-0.6B </summary>

```
vllm serve unsloth/Qwen3-0.6B-GGUF:IQ1_S --tokenizer Qwen/Qwen3-0.6B
INFO 11-21 10:25:52 [scheduler.py:207] Chunked prefill is enabled with max_num_batched_tokens=2048.
(APIServer pid=1206292) INFO 11-21 10:25:52 [api_server.py:1978] vLLM API server version 0.11.2.dev67+g1d642872a.d20251120
(APIServer pid=1206292) INFO 11-21 10:25:52 [utils.py:253] non-default args: {'model_tag': 'unsloth/Qwen3-0.6B-GGUF:IQ1_S', 'model': 'unsloth/Qwen3-0.6B-GGUF:IQ1_S', 'tokenizer': 'Qwen/Qwen3-0.6B'}
(APIServer pid=1206292) INFO 11-21 10:25:52 [arg_utils.py:1375]
(APIServer pid=1206292) INFO 11-21 10:25:52 [arg_utils.py:1375]
(APIServer pid=1206292) INFO 11-21 10:25:52 [arg_utils.py:1375] After maybe_override_with_speculators: unsloth/Qwen3-0.6B-GGUF
(APIServer pid=1206292) INFO 11-21 10:25:53 [model.py:652] Resolved architecture: Qwen3ForCausalLM
(APIServer pid=1206292) INFO 11-21 10:25:53 [model.py:1783] Using max model len 40960
(APIServer pid=1206292) INFO 11-21 10:25:53 [scheduler.py:207] Chunked prefill is enabled with max_num_batched_tokens=8192.
(EngineCore_DP0 pid=1206446) INFO 11-21 10:25:58 [core.py:93] Initializing a V1 LLM engine (v0.11.2.dev67+g1d642872a.d20251120) with config: model='unsloth/Qwen3-0.6B-GGUF', speculative_config=None, tokenizer='Qwen/Qwen3-0.6B', skip_tokenizer_init=False, tokenizer_mode=auto, revision=None, tokenizer_revision=None, trust_remote_code=False, dtype=torch.bfloat16, max_seq_len=40960, download_dir=None, load_format=gguf, tensor_parallel_size=1, pipeline_parallel_size=1, data_parallel_size=1, disable_custom_all_reduce=False, quantization=gguf, enforce_eager=False, kv_cache_dtype=auto, device_config=cuda, structured_outputs_config=StructuredOutputsConfig(backend='auto', disable_fallback=False, disable_any_whitespace=False, disable_additional_properties=False, reasoning_parser='', reasoning_parser_plugin='', enable_in_reasoning=False), observability_config=ObservabilityConfig(show_hidden_metrics_for_version=None, otlp_traces_endpoint=None, collect_detailed_traces=None), seed=0, served_model_name=unsloth/Qwen3-0.6B-GGUF, enable_prefix_caching=True, enable_chunked_prefill=True, pooler_config=None, compilation_config={'level': None, 'mode': <CompilationMode.VLLM_COMPILE: 3>, 'debug_dump_path': None, 'cache_dir': '', 'compile_cache_save_format': 'binary', 'backend': 'inductor', 'custom_ops': ['none'], 'splitting_ops': ['vllm::unified_attention', 'vllm::unified_attention_with_output', 'vllm::unified_mla_attention', 'vllm::unified_mla_attention_with_output', 'vllm::mamba_mixer2', 'vllm::mamba_mixer', 'vllm::short_conv', 'vllm::linear_attention', 'vllm::plamo2_mamba_mixer', 'vllm::gdn_attention_core', 'vllm::kda_attention', 'vllm::sparse_attn_indexer'], 'compile_mm_encoder': False, 'use_inductor': None, 'compile_sizes': [], 'inductor_compile_config': {'enable_auto_functionalized_v2': False, 'combo_kernels': True, 'benchmark_combo_kernel': True}, 'inductor_passes': {}, 'cudagraph_mode': <CUDAGraphMode.FULL_AND_PIECEWISE: (2, 1)>, 'cudagraph_num_of_warmups': 1, 'cudagraph_capture_sizes': [1, 2, 4, 8, 16, 24, 32, 40, 48, 56, 64, 72, 80, 88, 96, 104, 112, 120, 128, 136, 144, 152, 160, 168, 176, 184, 192, 200, 208, 216, 224, 232, 240, 248, 256, 272, 288, 304, 320, 336, 352, 368, 384, 400, 416, 432, 448, 464, 480, 496, 512], 'cudagraph_copy_inputs': False, 'cudagraph_specialize_lora': True, 'use_inductor_graph_partition': False, 'pass_config': {}, 'max_cudagraph_capture_size': 512, 'local_cache_dir': None}
(EngineCore_DP0 pid=1206446) INFO 11-21 10:25:59 [parallel_state.py:1217] world_size=1 rank=0 local_rank=0 distributed_init_method=tcp://172.16.1.13:46217 backend=nccl
[Gloo] Rank 0 is connected to 0 peer ranks. Expected number of connected peer ranks is : 0
[Gloo] Rank 0 is connected to 0 peer ranks. Expected number of connected peer ranks is : 0
[Gloo] Rank 0 is connected to 0 peer ranks. Expected number of connected peer ranks is : 0
[Gloo] Rank 0 is connected to 0 peer ranks. Expected number of connected peer ranks is : 0
[Gloo] Rank 0 is connected to 0 peer ranks. Expected number of connected peer ranks is : 0
[Gloo] Rank 0 is connected to 0 peer ranks. Expected number of connected peer ranks is : 0
[Gloo] Rank 0 is connected to 0 peer ranks. Expected number of connected peer ranks is : 0
(EngineCore_DP0 pid=1206446) INFO 11-21 10:25:59 [parallel_state.py:1425] rank 0 in world size 1 is assigned as DP rank 0, PP rank 0, PCP rank 0, TP rank 0, EP rank 0
(EngineCore_DP0 pid=1206446) INFO 11-21 10:25:59 [gpu_model_runner.py:3260] Starting to load model unsloth/Qwen3-0.6B-GGUF...
(EngineCore_DP0 pid=1206446) INFO 11-21 10:26:09 [cuda.py:415] Valid backends: ['FLASH_ATTN', 'FLASHINFER', 'TRITON_ATTN', 'FLEX_ATTENTION']
(EngineCore_DP0 pid=1206446) INFO 11-21 10:26:09 [cuda.py:424] Using FLASH_ATTN backend.
(EngineCore_DP0 pid=1206446) INFO 11-21 10:26:14 [gpu_model_runner.py:3339] Model loading took 0.2154 GiB memory and 14.283840 seconds
(EngineCore_DP0 pid=1206446) INFO 11-21 10:26:17 [backends.py:648] Using cache directory: /home/name/.cache/vllm/torch_compile_cache/124ca0c67d/rank_0_0/backbone for vLLM's torch.compile
(EngineCore_DP0 pid=1206446) INFO 11-21 10:26:17 [backends.py:708] Dynamo bytecode transform time: 3.01 s
(EngineCore_DP0 pid=1206446) INFO 11-21 10:26:20 [backends.py:214] Directly load the compiled graph(s) for dynamic shape from the cache, took 3.049 s
(EngineCore_DP0 pid=1206446) INFO 11-21 10:26:21 [monitor.py:34] torch.compile takes 6.06 s in total
(EngineCore_DP0 pid=1206446) INFO 11-21 10:26:22 [gpu_worker.py:337] Available KV cache memory: 65.28 GiB
(EngineCore_DP0 pid=1206446) INFO 11-21 10:26:22 [kv_cache_utils.py:1234] GPU KV cache size: 611,136 tokens
(EngineCore_DP0 pid=1206446) INFO 11-21 10:26:22 [kv_cache_utils.py:1239] Maximum concurrency for 40,960 tokens per request: 14.92x
(EngineCore_DP0 pid=1206446) 2025-11-21 10:26:22,420 - INFO - autotuner.py:256 - flashinfer.jit: [Autotuner]: Autotuning process starts ...
(EngineCore_DP0 pid=1206446) 2025-11-21 10:26:22,433 - INFO - autotuner.py:262 - flashinfer.jit: [Autotuner]: Autotuning process ends
Capturing CUDA graphs (mixed prefill-decode, PIECEWISE): 100%|██| 51/51 [00:01<00:00, 31.87it/s]
Capturing CUDA graphs (decode, FULL): 100%|█████████████████████| 51/51 [00:01<00:00, 42.92it/s]
(EngineCore_DP0 pid=1206446) INFO 11-21 10:26:25 [gpu_model_runner.py:4245] Graph capturing finished in 3 secs, took 0.71 GiB
(EngineCore_DP0 pid=1206446) INFO 11-21 10:26:25 [core.py:253] init engine (profile, create kv cache, warmup model) took 11.47 seconds
(APIServer pid=1206292) INFO 11-21 10:26:27 [api_server.py:1726] Supported tasks: ['generate']
(APIServer pid=1206292) INFO 11-21 10:26:29 [api_server.py:2056] Starting vLLM API server 0 on http://0.0.0.0:8000
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:38] Available routes are:
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /openapi.json, Methods: HEAD, GET
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /docs, Methods: HEAD, GET
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /docs/oauth2-redirect, Methods: HEAD, GET
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /redoc, Methods: HEAD, GET
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /health, Methods: GET
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /load, Methods: GET
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /tokenize, Methods: POST
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /detokenize, Methods: POST
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /v1/models, Methods: GET
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /version, Methods: GET
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /v1/responses, Methods: POST
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /v1/responses/{response_id}, Methods: GET
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /v1/responses/{response_id}/cancel, Methods: POST
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /v1/messages, Methods: POST
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /v1/chat/completions, Methods: POST
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /v1/completions, Methods: POST
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /v1/embeddings, Methods: POST
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /pooling, Methods: POST
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /classify, Methods: POST
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /score, Methods: POST
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /v1/score, Methods: POST
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /v1/audio/transcriptions, Methods: POST
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /v1/audio/translations, Methods: POST
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /rerank, Methods: POST
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /v1/rerank, Methods: POST
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /v2/rerank, Methods: POST
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /scale_elastic_ep, Methods: POST
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /is_scaling_elastic_ep, Methods: POST
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /inference/v1/generate, Methods: POST
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /ping, Methods: GET
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /ping, Methods: POST
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /invocations, Methods: POST
(APIServer pid=1206292) INFO 11-21 10:26:29 [launcher.py:46] Route: /metrics, Methods: GET
(APIServer pid=1206292) INFO:     Started server process [1206292]
(APIServer pid=1206292) INFO:     Waiting for application startup.
(APIServer pid=1206292) INFO:     Application startup complete.
```
</details>

<details>
<summary> pytest tests/models/test_gguf_download.py </summary>

```
pytest tests/models/test_gguf_download.py
===================================== test session starts ======================================
platform linux -- Python 3.12.3, pytest-9.0.1, pluggy-1.6.0
rootdir: /home/name/.test/vllm
configfile: pyproject.toml
plugins: anyio-4.11.0
collected 12 items

tests/models/test_gguf_download.py ............                                          [100%]

======================================= warnings summary =======================================
<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================ 12 passed, 2 warnings in 3.40s ================================
sys:1: DeprecationWarning: builtin type swigvarlink has no __module__ attribute
```
</details>

<details>
<summary> pytest tests/transformers_utils/test_utils.py </summary>

```
pytest tests/transformers_utils/test_utils.py
===================================== test session starts ======================================
platform linux -- Python 3.12.3, pytest-9.0.1, pluggy-1.6.0
rootdir: /home/name/.test/vllm
configfile: pyproject.toml
plugins: anyio-4.11.0
collected 12 items

tests/transformers_utils/test_utils.py ............                                      [100%]

======================================= warnings summary =======================================
<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute

<frozen importlib._bootstrap>:488
  <frozen importlib._bootstrap>:488: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================ 12 passed, 2 warnings in 1.78s ================================
sys:1: DeprecationWarning: builtin type swigvarlink has no __module__ attribute
```
</details>

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

injaeryou committed:
feat: hf_model_id:quant_type gguf

Signed-off-by: Injae Ryou <injaeryou@gmail.com>

injaeryou committed:
feat: add test

Signed-off-by: Injae Ryou <injaeryou@gmail.com>

sts07142 commented on vllm/engine/arg_utils.py:
```diff
@@ -1352,6 +1356,11 @@ def create_engine_config(
+        # Handle repo_id:quant_type for GGUF models
+        if is_remote_gguf(self.model):
+            repo_id, self.gguf_quant_type = self.model.rsplit(":", 1)
+            self.model = repo_id
```
@codex review
How about this?

def is_remote_gguf(model: str) -> bool:
    """Check if the model is a remote GGUF model."""
    return (
        (not is_cloud_storage(model))
        and (not model.startswith(("http://", "https://")))
        and ("/" in model and ":" in model)
    )

sts07142 replied:
Done

injaeryou committed:
fix: enhance remote GGUF model check logic

Signed-off-by: Injae Ryou <injaeryou@gmail.com>

sts07142 commented on vllm/model_executor/model_loader/gguf_loader.py:
```diff
@@ -55,12 +66,74 @@ def _prepare_weights(self, model_name_or_path: str):
+            )
+            local_files = [
+                f for f in all_gguf_files if quant_type in os.path.basename(f)
+            ]
```
I think the HF GGUF model follows the format below.
allow_patterns = [
            f"*-{quant_type}.gguf",
            f"*-{quant_type}-*.gguf",
            f"*/*-{quant_type}.gguf",
            f"*/*-{quant_type}-*.gguf",
        ]

injaeryou committed:
fix: 'is_gguf' to avoid AttributeError

Signed-off-by: Injae Ryou <injaeryou@gmail.com>

injaeryou committed:
refactor: update GGUF model loading methods to accept ModelConfig directly

- Changed `_prepare_weights` to take `ModelConfig` instead of a model path.
- Updated `download_gguf` to include an optional `revision` parameter.
- Adjusted `download_model` and `load_weights` methods to work with the new `_prepare_weights` signature.

Signed-off-by: Injae Ryou <injaeryou@gmail.com>

injaeryou committed:
refactor: remove fallback logic for GGUF file search

Signed-off-by: Injae Ryou <injaeryou@gmail.com>

Isotr0py reviewed (commented):
Thanks for adding this feature! I just leave some initial comments. PTAL! :)

hmellor commented on vllm/config/model.py:
```diff
@@ -113,6 +116,8 @@ class ModelConfig:
     content for `model_name` tag in metrics output when `served_model_name` is
     not specified."""
+    gguf_quant_type: str | None = None
+    """The quant type of the GGUF model to use."""
```
`QuantizationConfig` maybe?

injaeryou committed:
refactor: apply reviews

- remove 'gguf_quant_type' in ModelConfig
- move 'download_gguf' to weight_utils.py
- strictly check 'quant_type' in 'is_remote_gguf'
- leave self.model as repo_id:quant_type
- raise error in 'split_remote_gguf'
  - split invalid remote_gguf_model
  - invalid gguf_quant_type (different from GGMLQuantizationType)

Signed-off-by: Injae Ryou <injaeryou@gmail.com>

injaeryou committed:
refactor: tests

Signed-off-by: Injae Ryou <injaeryou@gmail.com>

sts07142 commented on vllm/model_executor/model_loader/gguf_loader.py:
```diff
@@ -55,12 +67,66 @@ def _prepare_weights(self, model_name_or_path: str):
+        for pattern in allow_patterns:
+            # Convert pattern to glob pattern for local filesystem
+            glob_pattern = os.path.join(folder, pattern)
+            local_files.extend(glob.glob(glob_pattern))
```
Done

Isotr0py committed:
better error message for missing config.json in model repo

Signed-off-by: Isotr0py <mozf@mail2.sysu.edu.cn>

Isotr0py commented on vllm/transformers_utils/config.py:
```diff
@@ -648,21 +653,44 @@ def get_config(
+                    "in engine args to fetch config from unquantized hf model."
+                )
+                logger.error(err_msg)
+                raise ValueError(err_msg)
```
I found that not each GGUF repo has `config.json` (`Qwen3-0.6-GGUF` etc), so I added an error message to ask user passing `--hf-config-path` manually as temporary solution.

Otherwise the error message from transformers is quite confusing:
```
(APIServer pid=1807)     s.__pydantic_validator__.validate_python(ArgsKwargs(args, kwargs), self_instance=s)
(APIServer pid=1807) pydantic_core._pydantic_core.ValidationError: 1 validation error for ModelConfig
(APIServer pid=1807)   Value error, Unrecognized model in Qwen/Qwen3-0.6B-GGUF. Should have a `model_type` key in its config.json, or contain one of the following strings in its name: aimv2, aimv2_vision_model, albert, align, altclip, apertus, arcee, aria, aria_text, audio-spectrogram-transformer, autoformer, aya_vision, bamba, bark, bart, beit, bert, bert-generation, big_bird, bigbird_pegasus, biogpt, bit, bitnet, blenderbot, blenderbot-small, blip, blip-2, blip_2_qformer, bloom, blt, bridgetower, bros, camembert, canine, chameleon, chinese_clip, chinese_clip_vision_model, clap, clip, clip_text_model, clip_vision_model, clipseg, clvp, code_llama, codegen, cohere, cohere2, cohere2_vision, colpali, colqwen2, conditional_detr, convbert, convnext, convnextv2, cpmant, csm, ctrl, cvt, d_fine, dab-detr, dac, data2vec-audio, data2vec-text, data2vec-vision, dbrx, deberta, deberta-v2, decision_transformer, deepseek_v2, deepseek_v3, deepseek_vl, deepseek_vl_hybrid, deformable_detr, deit, depth_anything, depth_pro, deta, detr, dia, diffllama, dinat, dinov2, dinov2_with_registers, dinov3_convnext, dinov3_vit, distilbert, doge, donut-swin, dots1, dpr, dpt, edgetam, edgetam_video, edgetam_vision_model, efficientformer, efficientloftr, efficientnet, electra, emu3, encodec, encoder-decoder, eomt, ernie, ernie4_5, ernie4_5_moe, ernie_m, esm, evolla, exaone4, falcon, falcon_h1, falcon_mamba, fastspeech2_conformer, fastspeech2_conformer_with_hifigan, flaubert, flava, flex_olmo, florence2, fnet, focalnet, fsmt, funnel, fuyu, gemma, gemma2, gemma3, gemma3_text, gemma3n, gemma3n_audio, gemma3n_text, gemma3n_vision, git, glm, glm4, glm4_moe, glm4v, glm4v_moe, glm4v_moe_text, glm4v_text, glpn, got_ocr2, gpt-sw3, gpt2, gpt_bigcode, gpt_neo, gpt_neox, gpt_neox_japanese, gpt_oss, gptj, gptsan-japanese, granite, granite_speech, granitemoe, granitemoehybrid, granitemoeshared, granitevision, graphormer, grounding-dino, groupvit, helium, hgnet_v2, hiera, hubert, hunyuan_v1_dense, hunyuan_v1_moe, ibert, idefics, idefics2, idefics3, idefics3_vision, ijepa, imagegpt, informer, instructblip, instructblipvideo, internvl, internvl_vision, jamba, janus, jetmoe, jukebox, kosmos-2, kosmos-2.5, kyutai_speech_to_text, layoutlm, layoutlmv2, layoutlmv3, led, levit, lfm2, lfm2_vl, lightglue, lilt, llama, llama4, llama4_text, llava, llava_next, llava_next_video, llava_onevision, longcat_flash, longformer, longt5, luke, lxmert, m2m_100, mamba, mamba2, marian, markuplm, mask2former, maskformer, maskformer-swin, mbart, mctct, mega, megatron-bert, metaclip_2, mgp-str, mimi, minimax, ministral, mistral, mistral3, mixtral, mlcd, mllama, mm-grounding-dino, mobilebert, mobilenet_v1, mobilenet_v2, mobilevit, mobilevitv2, modernbert, modernbert-decoder, moonshine, moshi, mpnet, mpt, mra, mt5, musicgen, musicgen_melody, mvp, nat, nemotron, nezha, nllb-moe, nougat, nystromformer, olmo, olmo2, olmo3, olmoe, omdet-turbo, oneformer, open-llama, openai-gpt, opt, ovis2, owlv2, owlvit, paligemma, parakeet_ctc, parakeet_encoder, patchtsmixer, patchtst, pegasus, pegasus_x, perceiver, perception_encoder, perception_lm, persimmon, phi, phi3, phi4_multimodal, phimoe, pix2struct, pixtral, plbart, poolformer, pop2piano, prompt_depth_anything, prophetnet, pvt, pvt_v2, qdqbert, qwen2, qwen2_5_omni, qwen2_5_vl, qwen2_5_vl_text, qwen2_audio, qwen2_audio_encoder, qwen2_moe, qwen2_vl, qwen2_vl_text, qwen3, qwen3_moe, qwen3_next, qwen3_omni_moe, qwen3_vl, qwen3_vl_moe, qwen3_vl_moe_text, qwen3_vl_text, rag, realm, recurrent_gemma, reformer, regnet, rembert, resnet, retribert, roberta, roberta-prelayernorm, roc_bert, roformer, rt_detr, rt_detr_resnet, rt_detr_v2, rwkv, sam, sam2, sam2_hiera_det_model, sam2_video, sam2_vision_model, sam_hq, sam_hq_vision_model, sam_vision_model, seamless_m4t, seamless_m4t_v2, seed_oss, segformer, seggpt, sew, sew-d, shieldgemma2, siglip, siglip2, siglip2_vision_model, siglip_vision_model, smollm3, smolvlm, smolvlm_vision, speech-encoder-decoder, speech_to_text, speech_to_text_2, speecht5, splinter, squeezebert, stablelm, starcoder2, superglue, superpoint, swiftformer, swin, swin2sr, swinv2, switch_transformers, t5, t5gemma, table-transformer, tapas, textnet, time_series_transformer, timesfm, timesformer, timm_backbone, timm_wrapper, trajectory_transformer, transfo-xl, trocr, tvlt, tvp, udop, umt5, unispeech, unispeech-sat, univnet, upernet, van, vaultgemma, video_llava, videomae, vilt, vipllava, vision-encoder-decoder, vision-text-dual-encoder, visual_bert, vit, vit_hybrid, vit_mae, vit_msn, vitdet, vitmatte, vitpose, vitpose_backbone, vits, vivit, vjepa2, voxtral, voxtral_encoder, wav2vec2, wav2vec2-bert, wav2vec2-conformer, wavlm, whisper, xclip, xcodec, xglm, xlm, xlm-prophetnet, xlm-roberta, xlm-roberta-xl, xlnet, xlstm, xmod, yolos, yoso, zamba, zamba2, zoedepth [type=value_error, input_value=ArgsKwargs((), {'model': ...rocessor_plugin': None}), input_type=ArgsKwargs]
```

Isotr0py commented on vllm/config/model.py:
```diff
@@ -969,13 +978,19 @@ def _verify_quantization(self) -> None:
-            # Check the text config as well for multi-modal models.
-            quant_cfg = self._parse_quant_hf_config(text_config)
+        if self.quantization == "gguf" or is_gguf(self.model):
+            quant_cfg = None
```
I think this is not very necessary. Because hf_config extracted from gguf model always won't have `"quantization_config"` field. In other words, `self._parse_quant_hf_config(self.hf_config)` will always return None for GGUF.

Isotr0py reviewed (approved):
Overall LGTM! Just leave a nit.

Isotr0py committed:
clean

Signed-off-by: Isotr0py <mozf@mail2.sysu.edu.cn>

Isotr0py committed:
Merge branch 'main' into feat/load-gguf-model

sts07142 commented:
> Nice!!! Thanks for the PR!!!
> 
> The `--tokenizer Qwen/Qwen3-0.6B` argument is required or optional?

The tokenizer argument is required for GGUF.
- We can use the original model repo as a tokenizer.

Almost all GGUF repo doesn't have tokenizer information.
So we need to use tokenizer argument.

ivanbaldo commented:
Strange, according to this diagram https://huggingface.co/docs/transformers/gguf the files can have the tokenizer embedded inside.
I wonder if vLLM already supports it in case it's there and so the --tokenizer option is optional.
